File: lib/dog-class.rb
###############################################################################
#
# dog-class.rb
#
# defines a class to hold information about dogs and store and load that
# information from an associated sql database. Current methods should be a load
# and store method. Maybe a seperate update method but probably not.
#
# February 1, 2017
# Erik Feller
#
################################################################################

class Dog
    
    #define data members of the class
    attr_accessor :id
    @id
    attr_accessor :name
    @name
    attr_accessor :breed_primary
    @breed_primary
    attr_accessor :breed_second
    @breed_second
    attr_accessor :gender
    @gender
    attr_accessor :age
    @age
    attr_accessor :status
    @status
    attr_accessor :intake_time
    @intake_time
    attr_accessor :hold_times
    @hold_times
    attr_accessor :out_time
    @out_time

    def load_data(id_query, db_handle)
        #Waiting to implement until I have db storing
    end

    def load_data_row(row)
        #A function that loads a dog from a DB row
        @id = row[0]
        @name = row[1]
        @breed_primary = row[2]
        @breed_second = row[3]
        @gender = row[4]
        @age = row[5]
        @status = row[6]
        @intake_time = row[7]
        @hold_times = row[8]
        @out_time = row[9]
        return self
    end

    def set(id, breed_primary, breed_second, gender, age, status, intake_time, out_time)
        #Change Attributes in case we need to
    end

    def store_data(db_handle)
        #Store the data of the current object 
        if is_new(db_handle)
           #puts("dog " + name + " is a new dog")
           db_handle.execute("INSERT INTO dogdata(id, name, breed_primary, breed_secondary, gender, age, status, in_time, hold_times, out_time) VALUES(?, ?, ?, ?, ?, ?, ?, ?, ?, ?)", [id, name, breed_primary, breed_second, gender, age, status, intake_time, hold_times, out_time])
        end
    end
    
    def update_data(db_handle)
      db_handle.execute("UPDATE dogdata SET id=?, name=?, breed_primary=?, breed_secondary=?, gender=?, age=?, status=?, in_time=?, hold_times=?, out_time=? WHERE id=?", [id, name, breed_primary, breed_second, gender, age, status, intake_time, hold_times, out_time, id])
    end

    def store_data_direct(id, breed_primary, breed_second, gender, age, status, intake_time)
        #Add data into the database
    end

    def is_new(db_handle)
      #determine if a dog is new to the database or not
      if db_handle.execute("SELECT id FROM dogdata WHERE id LIKE " + id.to_s).length == 0
        true
      else
        false
      end
    end

    def set_out_time()
      #Set the out time to the current time
      @out_time = Time.now.utc.to_i
    end

    def status_changed(db_loc)
      #determine if the status of a dog has been changed
    end

    def hold_time_record(db_loc)
      #record the specific times a dog has been placed on hold. Stored as comma seperated timestamps
    end
        
end

File: lib/dog_class.py
import datetime

class Dog:
    """A class to hold dog data and do some processing and storage"""

    def __init__(self, db_id, h_id, chip_num, name, p_breed, s_breed, age, gender, status, in_time, out_time):
        self.db_id = db_id
        self.h_id = h_id
        self.chip_num = chip_num
        self.name = name
        self.p_breed = p_breed
        self.s_breed = s_breed
        self.age = age
        self.gender = gender
        self.in_time = in_time 
        self.out_time = out_time
        #TODO:Add some bool values for behavior stuff
        
    def emptyDog():
        return Dog(None, None, None, None, None, None, None, None, None, None)


    def createFromAdoptableSearch(self, adoptableSearch):
        #later make a db query to see if the dog exists already
        self.h_id = int(adoptableSearch['ID'])
        if adoptableSearch['ChipNumber'] is not None:
            self.chip_num = int(adoptableSearch['ChipNumber'])
        else:
            self.chip_num = 0
        self.name = str(adoptableSearch['Name'])
        self.p_breed = str(adoptableSearch['PrimaryBreed'])
        self.s_breed = str(adoptableSearch['SecondaryBreed'])
        self.age = int(adoptableSearch['Age'])
        self.gender = str(adoptableSearch['Sex'])
        self.in_time = datetime.datetime.now()

    def createFromDbRow(self,row):
        if len(row) != 10:
            return "Format of database row has unexpectedly changed"
        else:
            self.db_id = row[0]
            self.h_id = row[1]
            self.name = row[2]
            self.age = row[3]
            self.p_breed = row[4]
            self.s_breed = row[5]
            self.gender = row[6]
            self.status = row[7]
            self.in_time = row[8]
            self.out_time = row[9]

    def updateInDb(self, conn):
        #Connect to database
        cursor = conn.cursor()
        #Check to see if the dog is already in the database
        cursor.execute("SELECT * FROM dogs WHERE humane_id=%s",(self.h_id,))
        rows = cursor.fetchall()
        if len(rows) > 0:
            #Dog is in database, just update
            #print("{0} is already in the database, updating their entry with the most recent info".format(self.name))
            cursor.execute("UPDATE dogs SET age=%s, breed_primary=%s, breed_secondary=%s, out_time=%s WHERE humane_id=%s", (self.age, self.p_breed, self.s_breed, self.out_time, self.h_id))
        else:
            #insert dog into database
            #print("Adding {0} to the database".format(self.name))
            #cursor.execute(self.sql_insert_print())
            cursor.execute("INSERT INTO dogs (humane_id, name, age, breed_primary, breed_secondary, gender, in_time, out_time) VALUES(%s, %s, %s, %s, %s, %s, %s, %s)", (self.h_id, self.name, self.age, self.p_breed, self.s_breed, self.gender, self.in_time, self.out_time))
        conn.commit()
        #print(cursor.fetchone())

    def setOutTime(self):
        self.out_time = datetime.datetime.now()

    #This function primarily serves a debug purpose
    def pretty_print(self):
        if self.s_breed == None:
            print(self.name + " is a " + self.p_breed + " that is " + str(self.age) + " months old " + self.gender)
        else:
            print(self.name + " is a " + self.p_breed + ", " + self.s_breed + " that is " + str(self.age) + " months old " + self.gender)

